build-system/tasks/helpers.js

```diff
@@ -374,96 +370,101 @@ function finishBundle(srcFilename, destDir, destFilename, options) {
-  function performBundle(continueOnError) {
-    let startTime;
-    return toPromise(
-      bundler
-        .bundle()
-        .once('readable', () => (startTime = Date.now()))
-        .on('error', (err) =>
-          handleBundleError(err, continueOnError, destFilename)
-        )
-        .pipe(source(srcFilename))
-        .pipe(buffer())
-        .pipe(sourcemaps.init({loadMaps: true}))
-        .pipe(
-          regexpSourcemaps(
-            /\$internalRuntimeVersion\$/g,
-            internalRuntimeVersion,
-            'runtime-version'
-          )
-        )
-        .pipe(regexpSourcemaps(/([^]+)/, devWrapper, 'wrapper'))
-        .pipe(rename(destFilename))
-        .pipe(sourcemaps.write('.'))
-        .pipe(gulp.dest(destDir))
-        .on('end', () =>
-          finishBundle(srcFilename, destDir, destFilename, options)
-        )
-    )
-      .then(() => {
-        let name = destFilename;
-        if (options.latestName) {
-          const latestMaxName = options.latestName.split('.js')[0] + '.max.js';
-          name = `${name} → ${latestMaxName}`;
-        }
-        endBuildStep('Compiled', name, startTime);
-      })
-      .then(() => {
-        const target = path.basename(destFilename, path.extname(destFilename));
-        if (UNMINIFIED_TARGETS.includes(target)) {
-          return applyAmpConfig(
-            `${destDir}/${destFilename}`,
-            /* localDev */ true,
-            /* fortesting */ options.fortesting
-          );
+  const esbuildBabelPlugin = {
+    name: 'babel',
+    async setup(build, {transform} = {}) {
+      const updateContents = (contents) => {
+        return contents
+          .replace(/\$internalRuntimeVersion\$/g, internalRuntimeVersion)
+          .replace(/([^]+)/g, wrapper.replace('<%= contents %>', '$1'));
+      };
+
+      const transformContents = async ({file, contents}) => {
+        const babelOptions = babel.loadOptions({
+          caller: {name: 'unminified'},
+          filename: file.path,
+          sourceFileName: path.relative(process.cwd(), file.path),
+        });
+        const result = await babel.transformAsync(contents, babelOptions);
+        return {contents: updateContents(result.code)};
+      };
+
+      if (transform) {
+        return await transformContents(transform);
+      }
+
+      build.onLoad({filter: /.*/, namespace: ''}, async (file) => {
+        if (cache.has(file.path)) {
+          return {contents: cache.get(file.path)};
         }
+        const contents = await fs.promises.readFile(file.path, 'utf-8');
+        const transformed = await transformContents({file, contents});
+        cache.set(file.path, transformed.contents);
+        return transformed;
       });
+    },
+  };
+
+  /**
+   * Bundles an entry point with all its imports and applies babel transforms.
+   * @param {boolean} continueOnError
+   */
+  async function performBundle(continueOnError) {
+    await esbuild
+      .build({
+        entryPoints: [entryPoint],
```

Comment:
I believe we can take advantage of both `watch` and `incremental` options

Reply:
The reason I don't think we should do this is that for extensions, the watcher needs to first build all CSS and then invoke the chain of `buildExtensionJs()` -> `compileUnminifiedJs()` -> `esbuild.build()`. I'm concerned that adding `watch` and `incremental` here will short-circuit the chain.

https://github.com/ampproject/amphtml/blob/11b1fcae9e6c7c7d829234e7a01f94033ea4dc72/build-system/tasks/extension-helpers.js#L435-L446

Okay to punt on this until I can get tests to work correctly, and attempt optimizations in a follow up PR?